File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    private static int N;
    private static int M;
    private static int[][] board = new int[505][505];
    private static boolean[][] visit = new boolean[505][505];
    private static int[] dx = {0, 0, 1, -1};
    private static int[] dy = {1, -1, 0, 0};
    private static int[][] exDx = {{0, 0, 0, 1}, {0, 0, 0, -1}, {0, 1, 2, 1}, {0, 1, 2, 1}};
    private static int[][] exDy = {{0, 1, 2, 1}, {0, 1, 2, 1}, {0, 0, 0, -1}, {0, 0, 0, 1}};
    private static int max = 0;

    // ㅜ 모양을 제외한 나머지 모양 탐색
    private static void dfs(int cur, int sum, int sx, int sy) {
        if (cur == 4) {
            max = Math.max(max, sum);
            return;
        }
        visit[sx][sy] = true;
        for (int i = 0; i < 4; i++) {
            int nx = sx + dx[i];
            int ny = sy + dy[i];
            if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
            if (visit[nx][ny]) continue;
            dfs(cur + 1, sum + board[nx][ny], nx, ny);
        }
        visit[sx][sy] = false;
    }

    // ㅜ 모양 탐색
    private static void exDfs(int sx, int sy) {
        for (int i = 0; i < 4; i++) {
            int sum = 0;
            for (int j = 0; j < 4; j++) {
                int nx = sx + exDx[i][j];
                int ny = sy + exDy[i][j];
                if (nx < 0 || nx >= N || ny < 0 || ny >= M) break;
                sum += board[nx][ny];
            }
            max = Math.max(max, sum);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }


        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                dfs(0, 0, i, j);
                exDfs(i, j);
            }
        }
        System.out.println(max);
    }
}

File: 백준/Gold/14500. 테트로미노/테트로미노.py
N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]

visit = [[False] * M for _ in range(N)]
dx = (-1, 1, 0, 0)
dy = (0, 0, -1, 1)

global answer
answer = 0
def dfs(x, y, sum_, cnt):
    global answer
    if cnt == 4:
        answer = max(answer, sum_)
        return
    visit[x][y] = True
    for d in range(4):
        nx = x + dx[d]
        ny = y + dy[d]
        if nx < 0 or nx >= N or ny < 0 or ny >= M: continue
        if visit[nx][ny]: continue
        dfs(nx, ny, sum_ + board[nx][ny], cnt + 1)
    visit[x][y] = False

def another_shape(x, y):
    global answer
    sum_ = board[x][y]
    directions = []
    for d in range(4):
        nx = x + dx[d]
        ny = y + dy[d]
        if nx < 0 or nx >= N or ny < 0 or ny >= M: continue
        directions.append(d)
        sum_ += board[nx][ny]
    
    if len(directions) <= 2:
        return
    if len(directions) == 3:
        answer = max(answer, sum_)
        return
    for d in directions:
        answer = max(answer, sum_ - board[x + dx[d]][y + dy[d]])


for i in range(N):
    for j in range(M):
        dfs(i, j, board[i][j], 1)
        another_shape(i, j)
print(answer)